Rework on Widgets

charlenni opened the pull request:
Second try. I now have made all changes in small parts with their own commits. I hope, I explained it all very well.

Run all test. They work without any error. I run WPF, MAUI, MapView and Blazor samples. Work. Found a bug in Blazor and removed it (last commit), which shouldn't have done here. Sorry.

charlenni commented:
There are still some changes for LoggingWidget to do (save all LogLevels, filter while drawing), but I will make a PR later.

pauldendulk commented:
Last week [this PR](https://github.com/Mapsui/Mapsui/pull/2370) with 92 files changed was closed because it was too big. Now you come back with this PR with 99 changes 😟.

I can look at the individual commits and perhaps understand the changes, but I will also have comments on those changes, and if requested changes are made there will be more comments and it will never really come to a conclusion. Unless I just decide to just merge because you have to merge at some point, but as I said in the other thread, that is not what I going to do anymore.

Also, you have to take me along with the reasoning behind these choices. I already see I made other choices in my branch working on the MapControl and perhaps there are good reasons for your choices but I have to understand the why. I am the maintainer of the project, I need to know why things are the way they are in order to understand how bugs should be solved and how new functionality should be added.

I am myself now working on the MapControl touch and inevitably on the widget touch logic changed in this PR. The bulk of the work is figuring out how things work, and then deciding how things should work, trying to come up with a clear responsibility for each component/method. 

Let's have a video chat session coming week to discuss how to proceed. I have time on Monday and Friday. You have my email address.

charlenni commented:
I'm sorry about the trouble. I know, that you are working on a total different part of the code, even if it is sometimes related.

You shouldn't look on the number of files changed. This number is getting high, if you change something, that is in many samples like the TextBoxWidget or the TextButtonWidget. Because of that, I tried to make as much commits as possible, so that the steps are very small and each step could be understand easily. There are only a few commits that contain more than one change.

Why I made this PR?
The overall idea behind this PR is to make life easier (as always 😉). I started with the integration of MapView to Mapsui. For me, MapView has to advantages over native Mapsui: First it has useful widgets already installed and it has an easy way to add pins. I think, that is what users want at the start with a new library. They want to have fast a nice map. But to add the useful widgets, I need the widgets (part of that is this PR) and the possibility to add it easy to the map (part of that is the MapBuilder/MapExtension discussion).

What I found?
I started long ago with the widgets. Now I found over the time, that the code for the widgets glowed, but not in a good way. There are many parts, that do the same but in a slightly different way. Much code, that could be reduced and made easier, more usable. E.g. the touchable interface. There was code in Widget to handle touches. Even widgets, that are not touchable implement this function. And already existing widgets could only with much effort made touchable. So I added a ITouchableWidget interface. Now you could made each widget touchable by implementing the interface. The difference between TextBoxWidget and TextButtonWidget is, that the later one implements ITouchableWidget interface. Same code for the widget, same code for the renderer. Even better, the renderer could be dropped, because now the function looks not only for a renderer for this type, but also for each base class. TextButtonWidget derives from TextBoxWidget and there is no renderer for TextButtonRenderer? No problem, use the renderer for TextBoxWidget. Is this set in MapRenderer? No. Another example: the arguments for a touch event. We had 3 of them. All slightly different, but nearly the same. HyperlinkTouchedEventArgs and WidgetTouchedEventArgs are the same but we have to look into both. I made one class for event args which works perfect. Or the IWidgetExtend interface which was introduced for widgets, that want to react on events for the whole screen. I removed this interface and added a enum for what events a widget like to have: only for the widget envelope or the whole view. If you now look into MouseCoordinatesWidget you will see a full working widget with the same functionality. It derives from TextBoxWidget and implements the ITouchableWidget interface. No need for calling Map.RefreshGraphics, because this is handled by changes of Text from TextBoxWidget. No need for use Map as argument for the constructor. Even no need for a special renderer, because it introduced nothing new over TextBoxWidget. I would say, MouseCoordinatesWidget is one of the best arguments for this PR.

Again, I'm sorry, that this PR is so big. But because there are some fundamental changes made, many parts of the existing code are involved. And I didn't see the possibility do make one after that before and you could review each of this by its own. You had each time remember again what we did last time and what is the bigger idea.

I started this PR a second time and now tried to make very small steps, that could be reviewed very easy. But I understand, that you missing the overall idea behind this. I try in the next comment to make a structure over all this commits, so that you get the idea behind them a little better. And yes, I could guide you through the changes. By text or while a meeting.

charlenni commented:
Ok, lets go in details.

I made small steps for each commit. Best would be to look on the commits one after one. The `Pref` and `Next` buttons in the right top are helpful for this. Don't look into the changed files. It is not very useful. Many different changes, that you couldn't follow because many different things are done at the same time. So you don't get the idea behind.

**Commit 1:** You saw this already. It is the float to double change.
**Commit Extend IWidget et al.:** Redesign of IWidget and all dependencies
**Commit Combine WidgetArgs et al.:** Combine WidgetTouchedEventArgs and WidgetArgs to a new WidgetTouchedEventArgs and replace all occurrences
**Commit Removed interfaces et al.:** Combined IWidgetTouchable and IWidgetExtended to a new ITouchableWidget interface
**Commit INotifyPropertyChanged et al.:** Implement INotifyPropertyChanged for Widget, so that all widgets can use it. Some had it already, some not. You prefer something else. No problem. Could be done very fast in Widget.
**Commit Change selection et al.:** Add the new TouchableArea to select, which widgets see the touch event.
**Commit Add stretch and absolute et al.:** Added two new alignments, so that a widget can fill the whole viewport (like LoggingWidget) or be set at an absolute position. Supported by a central function for calculating envelope in Widget.
**Commit Implement INotifyPropertyChanged et al.:** Updated many widgets to use the INotifyPropertyChanged interface. Always this is combined with new comments.
**Commit Add TextSize et al.:** Added TextSize to TextBoxWidget and remove it from the derived classes. Less code.
**Commit Rename WidgetTouched to Touched et al.:** No further information when you know that the touch event of a widget is a WidgetTouched.
**Commit Change image et al.:** Changed images for test. Some was wrong before. Some had errors in the code.
**Commit LoggingWidget full size et al.:** Implement new functions of widgets.
**Commit Move to folders et al.:** Sort the files into logical folders. Update usings.
**Commit Renamed ButtonWidget to IconButtonWidget et al.:** Devide ButtonWidget into IconButtonWidget and TextButtonWidget, because they are two different things.
**Commit Renamed to BoxWidgets et al.:** I decided to rename the folders from BoxWidget to BoxWidgets and so on to reflect that they contain more than one widget. Had to change many files for this.
**Commit Hyperlink et al.:** Renamed Hyperlink to HyperlinkWidget for consistency and derived from TextButtonWidget. Same to ZoomInOut. Huge code reduction.
**Commit Opacity to BoxWidget et al.:** Code reduction because many widgets derive from BoxWidget and many of them had an Opacity property.
**Commit Removed Touched et al.:** Made LoggingWidget untouchable.
**Commit Remove MouseCoordinatesWidgetRenderer et al.:** Added the possibility, that if no renderer is found, the renderer of the base class is used.
Commit Fast possibility MRect et al.: I changed MarginX/MarginY to a MRect Marging, PaddingX/PaddingY to a MRect Padding and PositionX/PositionY to a MPoint Position. Update of many files.

I'm sure in this way it should be doable in an hour. Thank you in advance for taking this time for an, what I think, useful rework of widgets.

inforithmics commented:
I had a quick overview of the pull request and I like the general direction.

1. The touch is now only handled in the ITouchableWidget good improvement I didn't do that because this would have been a breaking change.
2. Some Test Images are updated because the OSM copyright is now displayed.
3. Although the Widget hierarchy is no more logical.

inforithmics commented on Mapsui.Nts/Widgets/EditingWidget.cs:
```diff
@@ -3,43 +3,44 @@
     public EditManager EditManager { get; }
     public EditManipulation EditManipulation { get; }
 
+    public TouchableAreaType TouchableArea => TouchableAreaType.Widget;
```
good improvement, this avoids some workarounds like inflating the widget area to get the touch events on the the whole map control

inforithmics commented on Mapsui.Nts/Widgets/EditingWidget.cs:
```diff
@@ -3,43 +3,44 @@
     }
 
-    public bool HandleWidgetTouching(Navigator navigator, MPoint position, WidgetArgs args)
+    public bool HandleWidgetTouching(Navigator navigator, MPoint position, WidgetTouchedEventArgs args)
```
Better naming WidgetTouchedEventArgs than WidgetArgs

inforithmics commented on Mapsui.Nts/Widgets/EditingWidget.cs:
```diff
@@ -60,23 +61,19 @@ public bool HandleWidgetTouching(Navigator navigator, MPoint position, WidgetArg
+        var screenPosition = position;
 
-        if (EditManager.SelectMode)
+        if (args.LeftButton)
```
It is better to handle the mode on the arguments than global state which is always difficult to handle

inforithmics commented on Mapsui.Rendering.Skia/SkiaWidgets/BoxWidgetRenderer.cs:
```diff
@@ -10,19 +10,14 @@ public class BoxWidgetRenderer : ISkiaWidgetRenderer
-        backRect.Offset(offsetX, offsetY);
-        canvas.DrawRoundRect(backRect, boxWidget.CornerRadius, boxWidget.CornerRadius, backPaint);
-        boxWidget.Envelope = backRect.ToMRect();
+        canvas.DrawRoundRect(boxWidget.Envelope?.ToSkia() ?? new SKRect(0, 0, (float)boxWidget.Width, (float)boxWidget.Height), (float)boxWidget.CornerRadius, (float)boxWidget.CornerRadius, backPaint);
```
quite a simplification of code

inforithmics commented on Mapsui.Rendering.Skia/SkiaWidgets/LoggingWidgetRenderer.cs:
```diff
@@ -37,19 +37,26 @@ public void Draw(SKCanvas canvas, Viewport viewport, IWidget widget, float layer
-        var height = loggingWidget.Height;
-        var paddingX = loggingWidget.PaddingX;
-        var paddingY = loggingWidget.PaddingY;
+        var marginX = loggingWidget.Envelope.Left;
```
The Envelope makes the positioning of the Widgets easier than absolute distances

inforithmics commented on Mapsui.Rendering.Skia/SkiaWidgets/TextBoxWidgetRenderer.cs:
```diff
@@ -25,46 +27,38 @@ public static void DrawText(SKCanvas canvas, Viewport viewport, IWidget widget,
 
-        float paddingX = textBox.PaddingX;
-        float paddingY = textBox.PaddingY;
+        var paddingX = textBox.Padding.Left;
```
Padding is now Left and top instead of only x and y easier to unterstand

inforithmics commented on Mapsui.UI.MapView/MapView.cs:
```diff
@@ -779,21 +772,21 @@ private void UpdateButtonPositions()
         if (IsZoomButtonVisible)
         {
-            _mapZoomInButton!.Envelope = new MRect(newX, newY, newX + ButtonSize, newY + ButtonSize);
+            _mapZoomInButton!.Position = new MPoint(newX, newY);
```
With the position I don't have to calculate the Envelope my self easier usage

inforithmics commented on Mapsui.UI.Shared/MapControl.cs:
```diff
@@ -57,12 +58,10 @@ public partial class MapControl : INotifyPropertyChanged, IDisposable
     private ConcurrentQueue<IWidget>? _widgetCollection;
     // saving list of touchable Widgets
-    private List<IWidget>? _touchableWidgets;
+    private List<ITouchableWidget>? _touchableWidgets;
```
good improvement all touchable widgets implement ITouchableWidget

inforithmics commented on Mapsui.UI.Shared/MapControl.cs:
```diff
@@ -607,14 +605,16 @@ private void CommonDispose(bool disposing)
+        foreach (var widget in touchableWidgets)
         {
-            if (extendedWidget.HandleWidgetMoving(Map.Navigator, position, widgetArgs))
+            if (widget.HandleWidgetMoving(Map.Navigator, position, widgetArgs))
```
No need for extended widgets

inforithmics commented on Mapsui.UI.Shared/MapControl.cs:
```diff
@@ -734,9 +707,9 @@ private List<IWidget> GetTouchableWidgets()
             foreach (var widget in touchableWidgets)
             {
-                if (widget is IWidgetTouchable { Touchable: false }) continue;
+                if (widget is not ITouchableWidget) continue;
```
The interface says now if it is touchable and the touchable property is gone now

inforithmics commented on Mapsui/Layers/BaseLayer.cs:
```diff
@@ -176,7 +175,7 @@ protected set
 
     /// <inheritdoc />
-    public Hyperlink Attribution { get; set; } = new();
+    public HyperlinkWidget Attribution { get; set; } = new();
```
Improved Naming consistency All widgets should be called widgets.

inforithmics commented on Mapsui/Widgets/Alignment/HorizontalAlignment.cs:
```diff
@@ -4,6 +4,7 @@ public enum HorizontalAlignment
-    Right
-
+    Right,
+    Stretch,
```
Good improvement of Stretch alignment so now a widget can be programmed so that it fills the map automatically

inforithmics commented on Mapsui/Widgets/ButtonWidgets/IconButtonWidget.cs:
```diff
@@ -66,12 +95,12 @@ public class ButtonWidget : TextBox, INotifyPropertyChanged
     /// Rotation of the SVG image
     /// </summary>
-    public float Rotation
+    public double Rotation
```
everything double where possible

inforithmics commented on Mapsui/Widgets/IWidget.cs:
```diff
@@ -2,23 +2,43 @@
+    /// <summary>
+    /// Margin outside of the widget
+    /// </summary>
+    MRect Margin { get; set; }
```
Now the margin can be specified on all sides seperatly

inforithmics commented on Mapsui/Widgets/TouchableAreaType.cs:
```diff
@@ -0,0 +1,13 @@
+    /// <summary>
+    /// Whole Viewport is touchable
+    /// </summary>
+    Viewport
```
Good improvement now a widget can receive touch from the viewport, like the mousecoordinateswidget and Editwidgets

inforithmics commented on Mapsui/Widgets/Widget.cs:
```diff
@@ -1,51 +1,184 @@
 namespace Mapsui.Widgets;
 
-public abstract class Widget : IWidget, IWidgetTouchable
+public abstract class Widget : IWidget, INotifyPropertyChanged
```
IWidgetTouchable isn't anymore needed because only the interfaces that implment ITouchableWidget handle touch

inforithmics commented on Mapsui/Widgets/WidgetTouchedEventArgs.cs:
```diff
@@ -2,17 +2,41 @@
+    /// <summary>
+    /// Number of clicks on the Widget
+    /// </summary>
+    public int ClickCount { get; }
```
So more information is provided in the WidgetTouchetEventArgs simplifiying code

charlenni commented:
Thank you, @inforithmics, for your looking into this.

Could say something about the not logical widget hierarchy? Could be also in German if that is easier.

inforithmics commented:
I tested the Branch a little bit and found following behavior
When i open the logging widget
![grafik](https://github.com/Mapsui/Mapsui/assets/2290863/440d1179-f76f-43e3-9014-f33ea703b4c7)
Aftwards some remainder of the Logging widget is in every Map
![grafik](https://github.com/Mapsui/Mapsui/assets/2290863/456999b1-1f4e-44a7-8d3c-1d243adaf19d)

inforithmics commented:
> Thank you, @inforithmics, for your looking into this.
> 
> Could say something about the not logical widget hierarchy? Could be also in German if that is easier.

1. Previously all widgets had implemented a Touch Interface even if they didn't need it (for example TextboxWidget).
2. All widgets that need to render a Box are derived form the BoxWidget like TextBox, Button, previously more widget derived directly from the Widget class

charlenni commented:
> I tested the Branch a little bit and found following behavior
> 
> When i open the logging widget
> 
> ![grafik](https://github.com/Mapsui/Mapsui/assets/2290863/440d1179-f76f-43e3-9014-f33ea703b4c7)
> 
> Aftwards some remainder of the Logging widget is in every Map


This is normal behavior after Paul merged the PR #2380, where LoggingWidgets is automatically included when you in debug mode.

charlenni commented:
Super that you could take a look into this. I know, that it isn't very easy because of the number of changes, but I think it would be useful. And when I read your comments, I think it worked.

It would getting even better, if @janusw will add the possibility to have groups of widgets.

inforithmics commented:
> > I tested the Branch a little bit and found following behavior
> > When i open the logging widget
> > ![grafik](https://private-user-images.githubusercontent.com/2290863/294716536-440d1179-f76f-43e3-9014-f33ea703b4c7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ1ODAwOTcsIm5iZiI6MTcwNDU3OTc5NywicGF0aCI6Ii8yMjkwODYzLzI5NDcxNjUzNi00NDBkMTE3OS1mNzZmLTQzZTMtOTAxNC1mMzNlYTcwM2I0YzcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDZUMjIyMzE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDg2YmE3NjRiN2IyYTZkMjMwMzA4Nzg4ZDU3YTQ2MThlZDFhZDc3OTE3ZjcyN2ZhNTliNDQ1MWQyNWZjNTBkNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ko7xH32JwiE866Zs1m7uUxEb04Pt4h-aUmyOZFtDL3E)
> > Aftwards some remainder of the Logging widget is in every Map
> 
> This is normal behavior after Paul merged the PR #2380, where LoggingWidgets is automatically included when you in debug mode.

You are right I made a pull request that fixes this.
https://github.com/Mapsui/Mapsui/pull/2392

pauldendulk reviewed (commented):
Sorry I could not make time for this PR yet. I glanced through the code, I see a lot of good improvements, and some great (like reducing the number of renderers needed). So far I have two concerns. 

I am still hesitant about INotifyPropertyChanged in the widgets. We discussed this and decided to merge this including the INotifyPropertyChanged and will we revert it later if we choose some other mechanism.

This changes the widget event logic. This worries me because the current main branch has bugs in it. In the previous PR there we still many bugs that had to be merged from 4.1. I am working on this problem in this PR https://github.com/Mapsui/Mapsui/pull/2394. The reason it has bugs is that it is not clear how it should work. Our touch handing code is the messiest part of the project, and it is one of the main things to improve in v5. So, I would rather sit back and investigate, figure out how it works and then decide how it should work and only then change it. Then again if this PR introduces no fundamental changes in the widget events handling there is no problem. I want to investigate this by checking out this branch and to step though the event code. This is something I can not do before Friday. So, this will take more time.

pauldendulk reviewed (approved):
I checked out the branch, merged the conflicts and created a PR to charlennis branch. 

I did testing and code-stepping with the editing samples in WPF and Avalonia and they function as before. WPF works correct. Avalonia not. This will be dealt with in another PR.

I will merge this is my PR to this PR has been merged.
